File: app/src/main/java/com/nwe/spadesscore/PlayerNamesActivity.java
package com.nwe.spadesscore;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import androidx.appcompat.widget.SwitchCompat;
import androidx.core.content.ContextCompat;

import com.google.android.material.textfield.TextInputLayout;

import java.util.Objects;

/**
 * The PlayerNamesActivity class is responsible for managing the user interface
 * where players input their names before starting the Spades game.
 * <p>
 * This activity:
 * <ul>
 * <li>Validates player names to ensure they are not empty and do not exceed 10 characters.
 * <li>Displays warning dialogs if validation fails.
 * <li>Updates the SpadesGame singleton with the entered player names.
 * <li>Navigates to the DealCardsActivity to start the game after successful validation.
 * <li>Disables the back button to prevent unintended navigation.
 *  </ul>
 */
public class PlayerNamesActivity extends SpadesAppCompatActivity {

    private TextInputLayout player1_name_input, player2_name_input, player3_name_input, player4_name_input;

    private SwitchCompat random_dealer_switch;
    private final SpadesGame spadesGame = SpadesGame.getInstance();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    void initializeUIComponents() {
        player1_name_input = findViewById(R.id.player1_name_input);
        player2_name_input = findViewById(R.id.player2_name_input);
        player3_name_input = findViewById(R.id.player3_name_input);
        player4_name_input = findViewById(R.id.player4_name_input);
        random_dealer_switch = findViewById(R.id.random_dealer_switch);

        DEBUG_DELETE_LATER();
    }

    private void DEBUG_DELETE_LATER() {
        assert player1_name_input.getEditText() != null;
        player1_name_input.getEditText().setText("Player 1");
        assert player2_name_input.getEditText() != null;
        player2_name_input.getEditText().setText("Player 2");
        assert player3_name_input.getEditText() != null;
        player3_name_input.getEditText().setText("Player 3");
        assert player4_name_input.getEditText() != null;
        player4_name_input.getEditText().setText("Player 4");
    }

    @Override
    void setupUI() {
        if (spadesGame.getPlayerCount() == 3) {
            player4_name_input.setVisibility(View.GONE);
            findViewById(R.id.player4_space).setVisibility(View.GONE);
        }

        random_dealer_switch.setThumbTintList(ContextCompat.getColorStateList(this, R.color.switchThumbTint));
        random_dealer_switch.setTrackTintList(ContextCompat.getColorStateList(this, R.color.switchTrackTint));

        findViewById(R.id.start_game_button).setOnClickListener(v -> startGame());
    }

    @Override
    void initContentView() {
        setContentView(R.layout.activity_player_names);
    }

    private void startGame() {
        final String name_player1 = Objects.requireNonNull(player1_name_input.getEditText()).getText().toString();
        final String name_player2 = Objects.requireNonNull(player2_name_input.getEditText()).getText().toString();
        final String name_player3 = Objects.requireNonNull(player3_name_input.getEditText()).getText().toString();
        final String name_player4 = Objects.requireNonNull(player4_name_input.getEditText()).getText().toString();

        if (spadesGame.getPlayerCount() == 3) {
            if (isAnyNameEmpty(name_player1, name_player2, name_player3)
                    || isAnyNameTooLong(name_player1, name_player2, name_player3)) {
                return;
            } else {
                spadesGame.setPlayerNames(name_player1, name_player2, name_player3);
            }
        } else {
            if (isAnyNameEmpty(name_player1, name_player2, name_player3, name_player4)
                    || isAnyNameTooLong(name_player1, name_player2, name_player3, name_player4)) {
                return;
            } else {
                spadesGame.setPlayerNames(name_player1, name_player2, name_player3, name_player4);
            }
        }

        spadesGame.setRandomDealer(random_dealer_switch.isChecked());
        spadesGame.startGame();
        Intent switchToDealCardsActivity = new Intent(this, DealCardsActivity.class);
        startActivity(switchToDealCardsActivity);
    }

    private boolean isAnyNameEmpty(String... playerNames) {
        for (String playerName : playerNames) {
            if (playerName.trim().isEmpty()) {
                showEmptyNameWarningDialog();
                return true;
            }
        }
        return false;
    }

    private boolean isAnyNameTooLong(String... playerNames) {
        for (String playerName : playerNames) {
            if (playerName.length() > 10) {
                showTooLongNameWarningDialog();
                return true;
            }
        }
        return false;
    }

    private void showEmptyNameWarningDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage("Please give each player a name").setTitle("Empty Name");
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void showTooLongNameWarningDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage("Names can only be a maximum of 10 characters long").setTitle("Name too long");
        AlertDialog dialog = builder.create();
        dialog.show();
    }
}

File: app/src/main/java/com/nwe/spadesscore/PlayerNamesActivity.kt
package com.nwe.spadesscore

import android.app.AlertDialog
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.view.View
import android.widget.EditText
import androidx.activity.viewModels
import androidx.annotation.StringRes
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.google.android.material.color.MaterialColors
import com.google.android.material.switchmaterial.SwitchMaterial
import com.nwe.spadesscore.ui.gameRepository
import com.nwe.spadesscore.ui.names.PlayerNamesViewModel
import com.nwe.spadesscore.ui.names.StartGameResult

class PlayerNamesActivity : SpadesAppCompatActivity() {

    private val viewModel: PlayerNamesViewModel by viewModels {
        viewModelFactory { initializer { PlayerNamesViewModel(gameRepository) } }
    }

    private lateinit var player1NameInput: EditText
    private lateinit var player2NameInput: EditText
    private lateinit var player3NameInput: EditText
    private lateinit var player4NameInput: EditText
    private lateinit var randomDealerSwitch: SwitchMaterial

    override fun initContentView() {
        setContentView(R.layout.activity_player_names)
    }

    override fun initializeUIComponents() {
        player1NameInput = findViewById(R.id.player1_name_input)
        player2NameInput = findViewById(R.id.player2_name_input)
        player3NameInput = findViewById(R.id.player3_name_input)
        player4NameInput = findViewById(R.id.player4_name_input)
        randomDealerSwitch = findViewById(R.id.random_dealer_checkBox)
    }

    override fun setupUI() {
        if (viewModel.playerCount == 3) {
            findViewById<View>(R.id.player4_card).visibility = View.GONE
        }

        findViewById<View>(R.id.start_game_button).setOnClickListener { startGame() }

        applySwitchTints()
        applyFieldFocusListeners()
    }

    // ── Switch tints ─────────────────────────────────────────────────────────

    private fun applySwitchTints() {
        val accentColor = MaterialColors.getColor(randomDealerSwitch, R.attr.appAccent)
        val trackOffColor = MaterialColors.getColor(randomDealerSwitch, R.attr.appTrack)
        randomDealerSwitch.trackTintList = ColorStateList(
            arrayOf(intArrayOf(android.R.attr.state_checked), intArrayOf()),
            intArrayOf(accentColor, trackOffColor)
        )
        randomDealerSwitch.thumbTintList = ColorStateList.valueOf(Color.WHITE)
    }

    // ── Field focus highlight (bg_field ↔ bg_field_focus) ────────────────────

    private fun applyFieldFocusListeners() {
        val pairs = listOf(
            R.id.player1_card to player1NameInput,
            R.id.player2_card to player2NameInput,
            R.id.player3_card to player3NameInput,
            R.id.player4_card to player4NameInput
        )
        for ((cardId, editText) in pairs) {
            val card = findViewById<View>(cardId)
            editText.setOnFocusChangeListener { _, hasFocus ->
                card.setBackgroundResource(
                    if (hasFocus) R.drawable.bg_field_focus else R.drawable.bg_field
                )
            }
        }
    }

    // ── Text extraction ───────────────────────────────────────────────────────

    private fun nameOf(input: EditText): String = input.text?.toString().orEmpty()

    // ── Game start ────────────────────────────────────────────────────────────

    private fun startGame() {
        val names = buildList {
            add(nameOf(player1NameInput))
            add(nameOf(player2NameInput))
            add(nameOf(player3NameInput))
            if (viewModel.playerCount == 4) add(nameOf(player4NameInput))
        }
        when (viewModel.start(names, randomDealerSwitch.isChecked)) {
            StartGameResult.EmptyName ->
                showWarning(R.string.dialog_empty_name_title, R.string.dialog_empty_name_message)
            StartGameResult.NameTooLong ->
                showWarning(R.string.dialog_name_too_long_title, R.string.dialog_name_too_long_message)
            StartGameResult.Started ->
                startActivity(Intent(this, DealCardsActivity::class.java))
        }
    }

    private fun showWarning(@StringRes titleRes: Int, @StringRes messageRes: Int) {
        AlertDialog.Builder(this)
            .setTitle(titleRes)
            .setMessage(messageRes)
            .create()
            .show()
    }
}
